core/lib/spree/core/controller_helpers/order.rb

```diff
@@ -58,7 +58,22 @@ def associate_user
           session[:guest_token] = nil
         end
 
+        def remove_invalid_session_order_id
+          user = try_spree_current_user
+
+          if session[:order_id]
+            order = Spree::Order.find(session[:order_id])
+            if user && order.user != user
+              session[:order_id] = nil
+            elsif !user && && order.user_id
+              session[:order_id] = nil
+            end
+          end
+        end
```

Comment:
I would use a guard clause to avoid the deeply nested `if`. Also I think we can drop the nil checks on `user`, thus simplifying the conditional even further:

```ruby
def remove_invalid_session_order_id
  return unless session[:order_id]
  
  user = try_spree_current_user
  order = Spree::Order.find(session[:order_id])

  session[:order_id] = nil if order.user != user
end